pintracker/optracker/operation.go

```diff
@@ -160,16 +172,32 @@ func (op *Operation) Error() string {
 // an error message. It updates the timestamp.
 func (op *Operation) SetError(err error) {
 	_, span := trace.StartSpan(op.ctx, "optracker/SetError")
+
+	prevStatus := op.ToTrackerStatus()
+
 	op.mu.Lock()
 	{
 		op.phase = PhaseError
 		op.error = err.Error()
 		op.ts = time.Now()
 	}
 	op.mu.Unlock()
+
+	op.recordStatuses(prevStatus, op.ToTrackerStatus())
 	span.End()
 }
 
+func (op *Operation) recordStatuses(prevStatus api.TrackerStatus, newStatus api.TrackerStatus) {
+	if !prevStatus.Match(newStatus) {
+		stats.Record(op.ctx, observations.GetMeasureFromStatus(prevStatus).M(-1))
+		stats.Record(op.ctx, observations.GetMeasureFromStatus(newStatus).M(1))
+	}
+
+	if newStatus.Match(api.TrackerStatusUnpinned) {
+		stats.Record(op.ctx, observations.GetMeasureFromStatus(api.TrackerStatusPinned).M(-1))
+	}
+}
```

Comment:
If prevStatus is pinned. And new status is unpinned. Won't this will do M(-1) twice on the Pinned metric? Also, there is no Unpinned view anyway... the code seems a bit shaky here. Why not having a view for unpinned items?

Also, doing -1 on the old metrics and +1 in the new metrics also works if both are the same, so we can do that on every case and skip checking if they are the same or not.

Reply:
There are status change paths based on Operations. One is UnpinQueued to Unpinned, other is PinQueued to Pinned. Pinned and Unpinned marking end of an operation. So, even though a pinned item can be unpinned or vice-versa, phase changes will not happen from Pinned to Unpinned, because they are separate operation.

Reply:
> Why not having a view for unpinned items

Anything that is not in the shared state is unpinned. That way number of unpinned items are infinite.
We can have more lenient definition (only for the purpose of metrics), where we consider only those unpinned things that were pinned before. There are no states after unpinned. So, unpinned will always increase. If same item is pinned and unpinned multiple times, that would increase unpinned count.
Moreover, unlike other statuses, I fail to see value of measuring unpinned_count

If this convinces you I need to remove int64Measure meant to track this status.

Reply:
> Also, doing -1 on the old metrics and +1 in the new metrics also works if both are the same, so we can do that on every case and skip checking if they are the same or not.

Checking this condition reduce unnecessary function calls